### PR TITLE
Edit metadata redesign

### LINKED ISSUE #1722: Virtual metadata in edit metadata page
**Describe the bug**
The virtual relationship metadata on an item, that is added at run-time depending on the relationships that Item has with other entities, and thus not present in the DB, is present in the edit metadata page. This metadata is however not editable, since it is dependent on the relationships the item has, not actual metadata values on the item itself.

**To Reproduce**
1. Go to an item that has relationships to other entities
2. Go to the edit metadata page, eg https://demo7.dspace.org/entities/person/08f7d1b2-9bca-424e-a172-a6cd7eb943e2/edit/metadata
3. `relation.*` fiels are present, and appear editable
4. Edit one, change won't persist
5. Delete one, delete won't persist

**Expected behavior**
Since these types of metadata are not editable/delete-able directly (needs relationship changes to change) they should not appear on the `/edit/metadata` page or not _appear_ editable/delete-able (disabled on edit metadata page) since the changes won't persist.

### LINKED ISSUE #1894: Current item's metadata are cleared from display in the Edit item metadata page while editing
**Describe the bug**
While editing an item's metadata via the Edit item Metadata tab, occasionally all the metadata values are wiped off apart for the value currently being edited - see screenshot attached from the current DSpace7 DEMO site. 

I have replicated this behaviour both in my DSpace 7.4 updated environments and in the DEMO site, e.g. https://demo7.dspace.org/items/a1c5ed83-7bee-4e64-b66d-df3671d78459/edit/metadata

**To Reproduce**
1. As an Admin user, choose a random item and go to edit page
2. From the metadata tab, add a value and save
3. Then edit any value and after a couple of seconds the screen refreshes and no metadata values are shown

**Expected behavior**
All existing metadata are shown as expected

**Related work**
Link to any related tickets or PRs here.
Screenshot 2022-10-11 at 08 59 41

### PR DESCRIPTION
**Pull request by Atmire-Kristof**
## References
* Fixes #1722 
* Fixes #1894 

## Description
This PR redesigns the edit-metadata page for items (and adds support for other DSpaceObjects), solving several issues (see references)

## Instructions for Reviewers
Technical changes:
* Added a new module called `DsoSharedModule`, which can be imported wherever an edit metadata page is required for any DSpaceObject (and possible future shared components)
* Within this module, the main component for metadata edit is `DsoEditMetadataComponent` (`ds-themed-dso-edit-metadata` to support themes)
* `DsoEditMetadataForm` is a helper class dynamically storing information about the object's metadata and its changes made to them. It's also responsible for turning all the metadata changes into an array of Operations which can be sent to the REST API to update our object

Functionality changes:
* Metadata values are grouped by their field
* Metadata values are draggable within their field to change their place property
* Adding a new metadata value will automatically add it to the bottom of their respective field
* Metadata fields are sorted alphabetically

Testing guidelines:
* Edit an item and visit the "edit metadata" tab
* Click the add button to start adding a new value
* You should be able to search through all metadata fields using the field input
* Clicking the "confirm" button to the right of the value should add the new value at the bottom of their respective field and remain marked green
* Clicking the "confirm" button whilst an invalid metadata field is filled in, the user should get a warning underneath the input and the value isn't added
* Clicking "undo" (or "delete" on a confirmed value) should simply remove the new value
* Editing an existing metadata value should mark the value orange
* Marking an existing metadata value as "deleted" should mark the value red
* Moving a metadata value within its field should mark the values orange which's "place" properties are changed by the move
* After a move took place, the user should see an "undo" button below the respective metadata field and clicking it should undo any move changes within this field
* Clicking "discard" at the top or bottom of the table after any changes were made should get rid of these changes and revert the metadata table to its original state
* The "reinstate" button should be available after a discard took place and clicking it should bring back the changes that were previously discarded
* Clicking "save" should send a PATCH request to the REST api containing the changes made by the user (note: move changes are calculated dynamically and the operations might not reflect the order the user moved values in, but should have the same result regardless)
* A success notification should display once the object is successfully updated
* Editing a collection's item template should work the exact same as editing an item

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Review by atarix83 — CHANGES_REQUESTED** [review]
thanks @Atmire-Kristof for this PR

code looks overall good, and the refactoring works fine. I've just minor change requests to ask

- When the metadata has only one value, the drag button is disable, but the drag functionality is not.  
Dragging should be also disabled (you can use the cdkDragHandleDisabled property to do it), otherwise this is the effect : 
Schermata da 2023-01-19 18-14-49

- I'd increase the rows number of the textarea used for editing the value. Currently is set to 2 and it's a bit annoying to increase the size everytime when editing long fields. 
Schermata da 2023-01-19 18-19-04
I think 5 is a good compromise

**Comment by Atmire-Kristof** [comment]
@atarix83 Thank you for the feedback!

Those are some valid concerns, so I've fixed the dragging issue and increased the amount of rows on textareas to 5.

File: src/assets/i18n/en.json5 [hunk 1]
```diff
@@ -2460,6 +2470,64 @@
...
+
+  "itemtemplate.edit.metadata.edit.buttons.edit": "Edit",
+
+  "itemtemplate.edit.metadata.edit.buttons.remove": "Remove",
+
+  "itemtemplate.edit.metadata.edit.buttons.undo": "Undo changes",
+
+  "itemtemplate.edit.metadata.edit.buttons.unedit": "Stop editing",
+
+  "itemtemplate.edit.metadata.edit.buttons.virtual": "This is a virtual metadata value, i.e. a value inherited from a related entity. It can’t be modified directly. Add or remove the corresponding relationship in the \"Relationships\" tab",
```
Comment (tdonohue) [inline src/assets/i18n/en.json5:2489]: How can an Item Template have virtual metadata?  There's no "relationships" tab for a Template Item.

(If there's a good reason to leave this message, I'm OK with it.  Just pointing out that it seems unnecessary to me.)

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Atmire-Kristof : Overall, great work!  I gave this a review and test today.  Overall, it appears to work very well!  I'm not finding any obvious buggy behavior.  But, I did find some accessibility and styling issues.

First, the style issue.  It appears the style of all `ngbToolTip`s is changed on the Edit item page.  This seems accidental, as it affects also the tooltips in the breadcrumbs.

Here's the normal tooltip style on an Item page:
normal-tooltip-size

And, here's how that same tooltip looks on the "Edit Item" page (for that same Item):
large-tooltips


Second, while the features here are **great**.  This seems to have lost all the accessibility tags/elements that we had in the old "Edit Item" screen.  Most specifically, we stopped using `<table>` (which is problem for accessibility, see inline notes below), and all buttons are missing `title` attributes.  I think the buttons are an easy fix.  Not as sure about the `<table>` elements, but it'd be nice to move back to them if at all possible.

Overall though, great work.  This is a major improvement to the "Edit Item" screen.  It mainly needs better accessibility.  If somehow that's overly complex, we could create a follow-up ticket... but I'm hoping maybe you can simply replace some of the custom styling for `<div>` with actual table columns/cells.  From my understanding, using `<div>` for tables is problematic for accessibility.

**Comment by artlowel** [comment]
>Second, while the features here are great. This seems to have lost all the accessibility tags/elements that we had in the old "Edit Item" screen. Most specifically, we stopped using <table> (which is problem for accessibility, see inline notes below), and all buttons are missing title attributes. I think the buttons are an easy fix. Not as sure about the <table> elements, but it'd be nice to move back to them if at all possible.


@tdonohue We tried to do this using tables first, but we had to move away from them (and do all this work to refactor it to look the same using flex-box). When using tables we needed to use rowspan on the field name cells, which means the first row for a metadata field would have one cell more than the rows below. Not only did that make the html significantly more complicated, it led to all sorts of issues with drag-to-reorder, as you would be dragging rows of different sizes and outside of the context of the table (while dragging) rows wouldn't completely change their shape. So we can't go back to using `<table>` if we want to keep the grouping  by field, and drag to reorder

Instead I propose we add `role`  and `aria` attributes to the current design to make it as accessible as possible: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/table_role

**Comment by tdonohue** [comment]
@artlowel : Just to verify... Yes, I'm ok with using the `aria-role` attributes as a "workaround" to the fact that `<table>`s are hard to implement for this page.  I just want to be sure we are making this page as accessible as possible.

**Comment by artlowel** [comment]
@tdonohue Those roles and aria attributes have been added

**Review by atarix83 — APPROVED** [review]
thanks @Atmire-Kristof

LGTM now

**Review by tdonohue — APPROVED** [review]
👍 Thanks @Atmire-Kristof and @artlowel !  My prior feedback has been addressed & I see the new "role" attributes on the `<div>` tags representing the table, column, row, etc.   Also verified the small tooltip styling bug is fixed.  This all looks great now.

**Pull request merged**